libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3358,6 +3358,29 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 
 			annotation.isPure = isPure;
 		}
+	if (
+		auto const* functionType = dynamic_cast<FunctionType const*>(exprType);
+		!annotation.isPure.set() &&
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Event &&
+		functionType->hasDeclaration() &&
```

Comment:
Finally, since the value you're calculating is a boolean, you can just assign the result of your conditions to simplify it even further. E.g. for errors, after removing all the redundant checks:

```c++
	if (auto const* functionType = dynamic_cast<FunctionType const*>(exprType))
		annotation.isPure =
			functionType->kind() == FunctionType::Kind::Error &&
			memberName == "selector";
```